### PR TITLE
Add option to write test process stdout to file

### PR DESCRIPTION
**Pull request by evanweible-wf**
We've been using the workaround for #1625 (run build and tests separately) to collect the JSON output from the test runner, but it'd be nice for that to be supported in build_runner directly so that CI test configs are more straightforward.

This is an attempt at fixing #1625 by adding a `--test-stdout` option that, when set, results in the test command running the test process in normal mode and piping the stdout to the given file path. Stderr is still forwarded to the main build process so that non-test errors remain visible.

If this approach looks alright, I can follow up with a couple tests and changelog entry.

@jakemac53 @natebosch

### CODE REVIEW

File: build_runner/lib/src/entrypoint/test.dart [hunk 1]
```diff
@@ -17,6 +17,8 @@ import '../generate/build.dart';
 import 'base_command.dart';
 import 'options.dart';
 
+const _testStdout = 'test-stdout';
```
Comment (evanweible-wf) [inline build_runner/lib/src/entrypoint/test.dart]: Open to better name suggestions here

File: build_runner/lib/src/entrypoint/test.dart [hunk 2]
```diff
@@ -136,8 +146,30 @@ class TestCommand extends BuildRunnerCommand {
...
+        if (testStdout != null) {
+          var sink = testStdout.openWrite();
+          futures.add(sink
+              .addStream(testProcess.stdout)
+              .then((_) => sink.flush)
+              .then((_) => sink.close));
+        }
+        await Future.wait(futures);
+        stdout.writeln('Test results written to ${testStdout.path}');
+      } catch (e) {
```
Comment (natebosch) [inline build_runner/lib/src/entrypoint/test.dart]: What types of exceptions might we catch here?

**Comment by evanweible-wf** [comment]
@natebosch I like that approach better, too. Looking at the reporter implementations, I don't think it would be too difficult to add that in.

The `JsonReporter` funnels all of its output through an `_emit()` method that calls `print()`, so all of the logic in that reporter should be able to be shared with a file-based version that just writes to a file instead of printing.

The `CompactReporter` and `ExpandedReporter` don't centralize their printing anywhere (each have at least 7 calls to `print()`, but that should be easy enough to DRY up for the purposes of sharing with a file reporter. The `CompactReporter` also uses `stdout.write` since it rewrites lines, but those could be ignored when writing to a file – presumably if someone wants to write the compact output to file, all they care about is the final result in a single line.

The runner does have some special casing when in debug mode based on the reporter – it will only output the debug info to the terminal if it's not the json reporter so as to not break its contract, but that wouldn't apply to the file reporters.

Since the `Runner` constructor is private, I think we could update it to accept a file reporter as an optional param without it being a breaking change, and then any place where the reporter is interacted with would just need to be updated to also interact with the file reporter.

Alternatively, the runner and `debug()` fn could be updated to accept a list of reporters and then file-based reporters could be added to the existing reporter registry, but that seems a bit odd.. you would never enable multiple stdout reporters and I can't think of a scenario where multiple file reporters would be useful.

File: build_runner/lib/src/entrypoint/test.dart (hunk 2 above)
Comment (evanweible-wf) [reply build_runner/lib/src/entrypoint/test.dart]: Probably just `FileSystemException`s, potentially when creating the file. I ended up having to update line 160 locally to support writing to a directory that doesn't yet exist:

```
var sink = (testStdout..createSync(recursive: true)).openWrite();
```

**Comment by natebosch** [comment]
I got started with https://github.com/dart-lang/test/pull/1101 to experiment.
As far as I can tell it should be safe to have multiple reporters and to allow a reporter to write to an arbitrary `Sink<List<int>>`

We'll need to figure out how to handle option parsing for this, but I think it's promising.

**Comment by evanweible-wf** [comment]
I'm going to close this since it seems to make a lot more sense in pkg:test.

**Pull request closed without merging**